### ISSUE TITLE
Support for Custom Value Serializer in Log4jEventBuilder#addKeyValue() and Asynchronous Serialization

### ISSUE THREAD

**Issue by AlexxeyKa** [apache/logging-log4j2#2996]
Hello all,

Thank you for providing detailed comments [here](https://github.com/apache/logging-log4j2/issues/1414#issuecomment-1748852847).

The mechanism introduced in `SLF4J` 2.x with the `addKeyValue` method looks very promising, but I found a couple of issues preventing my team from fully utilizing it:

1. Complex Objects Serialization: The current implementation relies on `String.valueOf(value)`, which is insufficient for handling complex objects, especially when structured logging is required.
2. Performance: The serialization happens synchronously, which can degrade performance in high-performance systems where complex objects need to be logged frequently.

I would like to request an enhancement in Log4j that allows users to set a custom serializer.
In the current `Log4jEventBuilder#addKeyValue(String key, Object value)` implementation, complex objects are serialized using `String.valueOf()`, which lacks flexibility. My suggestion is to allow users to provide a custom serializer that can handle more complex serialization needs (e.g., converting objects to `JSON`).

Current Implementation in Log4j:
```
@Override
public LoggingEventBuilder addKeyValue(final String key, final Object value) {
    if (keyValuePairs == null) {
        keyValuePairs = new HashMap<>();
    }
    keyValuePairs.put(key, String.valueOf(value));  // Uses String.valueOf(value)
    return this;
}
```

Additionally, I propose that the serialization of the value should happen asynchronously, offloading the potentially expensive serialization task to a separate thread. This will prevent blocking the thread that calls logging methods (such as `log.info()`), thus improving performance in high-throughput logging environments where structured logging is used extensively.

Thank you for your consideration.

**Comment by rgoers** [comment]
LOL.  @vy Remember the Renderer I wanted to add for use in ScopedContext? Here is another use for it.

**Comment by vy** [comment]
> Remember the Renderer I wanted to add for use in ScopedContext?

@rgoers, yes, I do. Note that I wasn't against the `Renderer` concept, but introducing a new interface where existing ones (e.g., `StringBuilderFormattable`, `MultiFormatStringBuilderFormattable`) can perfectly suffice. See [my comment](/apache/logging-log4j2/pull/2385#discussion_r1531746011).

> Here is another use for it.

User is using SLF4J. You're suggesting him to

1. either completely switch to the Log4j API and use `ScopedContext`<sup>1</sup>
2. or combine SLF4J with Log4j API's `ScopedContext`<sup>1</sup> (hence, effectively use both logging APIs).

@AlexxeyKa will be the judge of this.

<sup>1</sup> `ScopedContext` will be released in the future as a separate module to enhance Log4j API with a scoped thread context pretty much identical to the one you would get with Java's `ScopedValue`.

**Comment by ppkarwasz** [comment]
It is worth noting that in #1813, we agreed to that `addKeyValue()` should generate a [`MapMessage`](https://logging.apache.org/log4j/2.x/javadoc/log4j-api/org/apache/logging/log4j/message/MapMessage.html) instead of adding keys to the context data of the log event.

For those not accustomed to the difference between log message and event, there is a note in the [Messages documentation page](https://logging.apache.org/log4j/2.x/manual/messages.html).
Currently `addKeyValue()` fills the content of "labels", whereas it should fill the content of "message" in the example below:

```json
{

  "log.level": "INFO",
  "message": "Unable to insert data into my_table.",
  "error.type": "java.lang.RuntimeException",
  "error.message": null,
  "error.stack_trace": [
    {
      "class": "com.example.Main",
      "method": "doQuery",
      "file.name": "Main.java",
      "file.line": 36
    },
    {
      "class": "com.example.Main",
      "method": "main",
      "file.name": "Main.java",
      "file.line": 25
    }
  ],
  "marker": "SQL",
  "log.logger": "com.example.Main",

  "tags": [
    "SQL query"
  ],
  "labels": {
    "span_id": "3df85580-f001-4fb2-9e6e-3066ed6ddbb1",
    "trace_id": "1b1f8fc9-1a0c-47b0-a06f-af3c1dd1edf9"
  },

  "@timestamp": "2024-05-23T09:32:24.163Z",
  "log.origin.class": "com.example.Main",
  "log.origin.method": "doQuery",
  "log.origin.file.name": "Main.java",
  "log.origin.file.line": 36,
  "process.thread.id": 1,
  "process.thread.name": "main",
  "process.thread.priority": 5
}
```

**Comment by AlexxeyKa** [comment]
@ppkarwasz Thank you for noticing this.

If the map with key-values passed via `addKeyValue()` remains as `Map<String, Object>` in the `Message` (and values will not be stringified) it should solve my request.

**Comment by plesingr-aeb** [comment]
I face the same issue. I agree with @AlexxeyKa and would prefer if `Log4jEventBuilder` sticks to the `addKeyValue(String key, Object value)` API and does not deliberately transforms the objects to Strings.
